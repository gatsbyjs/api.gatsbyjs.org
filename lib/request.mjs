import axios from 'axios';
import getLogger from './logger.mjs';

const logger = getLogger('lib/request');

const logAndReturn = val => {
  const { status, statusText, data } = val;
  logger.verbose('%d: %s', status, statusText);

  return val;
};

export const put = (uri, payload, config = {}) => {
  logger.verbose('Sending a PUT request to %s', uri);
  logger.verbose('Payload:', payload);
  return axios.put(uri, payload, config).then(logAndReturn);
};

export const post = (uri, payload, config = {}) => {
  logger.verbose('Sending a POST request to %s', uri);
  logger.verbose('Payload:', payload);
  return axios.post(uri, payload, config).then(logAndReturn);
};
