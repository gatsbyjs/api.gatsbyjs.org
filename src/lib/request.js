import axios from 'axios';
import querystring from 'querystring';
import getLogger from './logger';

const logger = getLogger('lib/request');

export const put = (uri, payload, config = {}) => {
  logger.verbose('Sending a PUT request to %s', uri);
  return axios.put(uri, payload, config);
};

export const post = (uri, payload, config = {}) => {
  logger.verbose('Sending a POST request to %s', uri);
  return axios.post(uri, payload, config);
};

export const get = (uri, queryObject, config = {}) => {
  const queryString = querystring.stringify(queryObject);
  logger.verbose('Sending a GET request to %s', uri);
  logger.verbose('QueryString: %s', queryString);

  return axios.get(`${uri}?${queryString}`, config);
};
