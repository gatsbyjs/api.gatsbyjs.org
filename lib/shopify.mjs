import dotenv from 'dotenv';
import { post } from './request.mjs';
import getLogger from './logger.mjs';

dotenv.config();

const logger = getLogger('lib/shopify');

const verifyCustomerExists = error => {
  const { response: { data = {}, status, statusText } = {} } = error;
  logger.verbose('%d: %s — %s', status, statusText, JSON.stringify(data));

  // If the error is that the customer already exists, we can safely ignore that.
  if (
    data.errors &&
    data.errors.email &&
    data.errors.email[0] === 'has already been taken'
  ) {
    return true;
  }

  // If we get here, something went wrong that we didn’t expect.
  logger.error('Error %d: %s', status, statusText, { data, error });

  return false;
};

const sendPost = (endpoint, payload) => {
  const key = process.env.SHOPIFY_API_KEY;
  const secret = process.env.SHOPIFY_API_SECRET;
  const domain = process.env.SHOPIFY_URI;

  const uri = `https://${key}:${secret}@${domain}${endpoint}`;

  return post(uri, payload);
};

export const createShopifyCustomer = async ({
  username,
  email,
  first_name,
  subscribe
}) => {
  const endpoint = '/admin/customers.json';
  const payload = {
    customer: {
      first_name,
      email,
      accepts_marketing: subscribe,
      tags: 'contributor',
      metafields: [
        {
          key: 'github',
          value: username,
          namespace: 'global',
          value_type: 'string'
        }
      ]
    }
  };

  let customer = false;
  let errorMessage;
  try {
    logger.verbose('Sending POST request to Shopify...');
    const response = await sendPost(endpoint, payload);
    logger.verbose('POST response:');
    logger.verbose(JSON.stringify(response.data));

    if (response.status === 201) {
      logger.verbose('@%s was added as a customer', username);
      customer = true;
    }

    if (!customer) {
      logger.verbose(
        'There was a problem adding @%s as a customer: %s',
        username,
        response.statusText
      );
      logger.error(JSON.stringify(response));
      errorMessage = response.statusText;
    }
  } catch (error) {
    logger.verbose('An error was thrown while trying to create a customer.');

    // If the customer was already created in Shopify, we get an error back.
    // However, we don’t want to fail for that case, so check for it here.
    customer = verifyCustomerExists(error);

    logger.verbose(
      '@%s %s already a customer',
      username,
      customer ? 'is' : 'is not'
    );

    if (!customer) {
      errorMessage = 'Unable to register the user with Shopify';
    }
  }

  if (errorMessage) {
    logger.error(errorMessage);
    throw new Error(errorMessage);
  }

  return customer;
};
