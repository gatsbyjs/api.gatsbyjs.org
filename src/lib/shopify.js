import { get, post, put } from './request';
import getLogger from './logger';

const logger = getLogger('lib/shopify');

const getShopifyEndpoint = endpoint => {
  const key = process.env.SHOPIFY_API_KEY;
  const secret = process.env.SHOPIFY_API_SECRET;
  const domain = process.env.SHOPIFY_URI;

  return `https://${key}:${secret}@${domain}${endpoint}`;
};

export const addContributorTagToCustomer = async customer => {
  const tags = customer.tags.split(',');

  if (tags.includes('contributor')) {
    return true;
  }

  const nextTags = ['contributor', ...tags].join();
  const uri = getShopifyEndpoint(`/admin/customers/${customer.id}.json`);
  const payload = { customer: { id: customer.id, tags: nextTags } };

  const response = await put(uri, payload);

  if (response.status !== 200) {
    throw new Error('Unable to add the “contributor” tag to the customer');
  }

  logger.verbose(
    'The “contributor” tag was added to the customer %s',
    payload.customer.email
  );

  return true;
};

export const getCustomerByEmail = async email => {
  const endpoint = `/admin/customers/search.json`;
  const uri = getShopifyEndpoint(endpoint);

  const { data: { customers = [] } = {} } = await get(uri, { query: email });

  if (!customers[0]) {
    throw new Error(`No customer found with the email ${email}`);
  }

  return customers[0];
};

export const createCustomer = async payload => {
  try {
    const uri = getShopifyEndpoint('/admin/customers.json');
    const response = await post(uri, payload);

    if (response.status !== 201) {
      throw new Error(response.statusText);
    }

    logger.verbose(
      '%s was added as a customer in Shopify',
      payload.customer.email
    );
    return true;
  } catch (e) {
    /*
     * If the attempt to create the customer fails, it means the customer
     * already exists, so we need to load the existing customer.
     */
    const customer = await getCustomerByEmail(payload.customer.email);

    // Ensure the customer has the `contributor` tag so the discount works.
    await addContributorTagToCustomer(customer);

    return true;
  }
};

export const createShopifyCustomer = async ({
  username,
  email,
  first_name,
  subscribe
}) => {
  const response = await createCustomer({
    customer: {
      first_name,
      email,
      accepts_marketing: subscribe,
      tags: 'contributor',
      metafields: [
        {
          key: 'github',
          value: username,
          namespace: 'global',
          value_type: 'string'
        }
      ]
    }
  });

  return response;
};
