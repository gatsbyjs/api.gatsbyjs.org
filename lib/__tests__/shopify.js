import * as request from '../request.mjs';
import { createShopifyCustomer } from '../shopify.mjs';

jest.mock('dotenv', () => ({
  config: () => {
    // Provide the env vars expected by the lib
    process.env.SHOPIFY_API_KEY = 'TEST_SHOPIFY_API_KEY';
    process.env.SHOPIFY_API_SECRET = 'TEST_SHOPIFY_API_SECRET';
    process.env.SHOPIFY_URI = 'TEST_SHOPIFY_URI';
    process.env.SHOPIFY_DISCOUNT_CODE = 'TEST_SHOPIFY_DISCOUNT_CODE';
  }
}));

let mockReturn = () => {};
const mockPayload = {
  username: 'testuser',
  email: 'test@example.org',
  first_name: 'Test',
  subscribe: true
};

jest.mock('../request.mjs', () => ({
  get: jest.fn(() => mockReturn),
  post: jest.fn(() => mockReturn),
  put: jest.fn(() => mockReturn)
}));

describe('lib/shopify', () => {
  describe('createShopifyCustomer()', () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    test('hits the expected endpoint with a well-formed payload', async () => {
      request.post.mockReturnValueOnce(Promise.resolve({ status: 201 }));
      await createShopifyCustomer(mockPayload);

      expect(request.post).toMatchSnapshot();
    });

    test('doesn’t sign the user up for emails if they didn’t subscribe', async () => {
      request.post.mockReturnValueOnce(Promise.resolve({ status: 201 }));
      await createShopifyCustomer({ ...mockPayload, subscribe: false });

      expect(request.post).toMatchSnapshot();
    });

    test('creates a new customer in Shopify', async () => {
      request.post.mockReturnValueOnce(Promise.resolve({ status: 201 }));

      const response = await createShopifyCustomer(mockPayload);

      expect(response).toBe(true);
    });

    test('doesn’t throw an error if the customer already exists', async () => {
      request.post.mockReturnValueOnce(
        Promise.reject({
          response: { data: { errors: { email: ['has already been taken'] } } }
        })
      );

      request.get.mockReturnValueOnce(
        Promise.resolve({
          data: { customers: [{ id: 1234, tags: 'contributor' }] }
        })
      );

      request.put.mockReturnValueOnce(
        Promise.resolve({
          status: 200
        })
      );

      const response = await createShopifyCustomer(mockPayload);
      expect(response).toBe(true);
    });

    test('throws an error if something other than a 201 comes back', () => {
      request.post.mockReturnValueOnce(
        Promise.resolve({ status: 404, statusText: 'Not Found' })
      );

      return expect(createShopifyCustomer(mockPayload)).rejects.toThrowError(
        'Not Found'
      );
    });

    test('throws an error if an empty response comes back', () => {
      request.post.mockReturnValueOnce(Promise.reject({}));
      request.get.mockReturnValueOnce(Promise.reject({}));

      return expect(createShopifyCustomer(mockPayload)).rejects.toThrowError(
        'Unable to retrieve customers.'
      );
    });
  });
});
