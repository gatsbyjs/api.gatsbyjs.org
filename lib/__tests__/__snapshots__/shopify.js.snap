// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/shopify createShopifyCustomer() doesn’t sign the user up for emails if they didn’t subscribe 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://TEST_SHOPIFY_API_KEY:TEST_SHOPIFY_API_SECRET@TEST_SHOPIFY_URI/admin/customers.json",
      Object {
        "customer": Object {
          "accepts_marketing": false,
          "email": "test@example.org",
          "first_name": "Test",
          "metafields": Array [
            Object {
              "key": "github",
              "namespace": "global",
              "value": "testuser",
              "value_type": "string",
            },
          ],
          "tags": "contributor",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`lib/shopify createShopifyCustomer() hits the expected endpoint with a well-formed payload 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://TEST_SHOPIFY_API_KEY:TEST_SHOPIFY_API_SECRET@TEST_SHOPIFY_URI/admin/customers.json",
      Object {
        "customer": Object {
          "accepts_marketing": true,
          "email": "test@example.org",
          "first_name": "Test",
          "metafields": Array [
            Object {
              "key": "github",
              "namespace": "global",
              "value": "testuser",
              "value_type": "string",
            },
          ],
          "tags": "contributor",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;
